src/java/com/palantir/cassandra/check/VolumeIntegrityCheck.java

```diff
@@ -0,0 +1,114 @@
+
+import java.io.IOException;
+import java.nio.file.Paths;
+import java.util.Arrays;
+import java.util.Optional;
+import java.util.UUID;
+
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.palantir.cassandra.utils.FileParser;
+import com.palantir.logsafe.Preconditions;
+import org.apache.cassandra.config.DatabaseDescriptor;
+
+/**
+ * Check if the backing volume has been tampered with. While Cassandra supports multiple data drive, we only
+ * use 1 data drive.
+ */
+public final class VolumeIntegrityCheck
+{
+    public static final String METADATA_NAME = "cassandra-metadata.json";
+
+    private final UUID hostId;
+
+    private final FileParser<VolumeMetadata> dataDriveMetadataFileParser;
+
+    private final FileParser<VolumeMetadata> commitLogMetadataFileParser;
+
+    static class VolumeMetadataType extends TypeReference<VolumeMetadata>
+    {
+    }
+
+    public VolumeIntegrityCheck(UUID hostId,
+                                FileParser<VolumeMetadata> dataDriveMetadataFileParser,
+                                FileParser<VolumeMetadata> commitLogMetadataFileParser)
+    {
+        this.hostId = hostId;
+        this.dataDriveMetadataFileParser = dataDriveMetadataFileParser;
+        this.commitLogMetadataFileParser = commitLogMetadataFileParser;
+    }
+
+    public void execute()
+    {
+        try
+        {
+            Optional<VolumeMetadata> maybeDataDriveMetadata = dataDriveMetadataFileParser.read();
+            Optional<VolumeMetadata> maybeCommitLogMetadata = commitLogMetadataFileParser.read();
+            check(maybeDataDriveMetadata, maybeCommitLogMetadata);
+
+            // Write order matters here. We treat write to the data drive as
+            // a "locking" operation, i.e the node is in a steady state.
+            writeIfEmpty(commitLogMetadataFileParser, maybeCommitLogMetadata.isPresent());
+            writeIfEmpty(dataDriveMetadataFileParser, maybeDataDriveMetadata.isPresent());
+        }
+        catch (IOException cause)
+        {
+            throw new RuntimeException("Fail to execute check for volume integrity", cause);
+        }
+    }
+
+    private void check(Optional<VolumeMetadata> maybeDataDriveMetadata, Optional<VolumeMetadata> maybeCommitLogMetadata)
+    {
+        if (maybeCommitLogMetadata.isPresent())
+        {
+            Preconditions.checkState(VolumeMetadata.of(hostId).equals(maybeCommitLogMetadata.get()),
+                                     "Expected SystemKeyspace HostId and POD_NAME to match with commitlog");
+        }
+        else
+        {
+            Preconditions.checkState(!maybeDataDriveMetadata.isPresent(),
+                                     "Expected no metadata has been written to data drive if commitlog is empty");
+        }
+    }
+
+    private void writeIfEmpty(FileParser<VolumeMetadata> parser, boolean present) throws IOException
+    {
+        if (!present)
+        {
+            parser.write(VolumeMetadata.of(hostId));
+        }
```

Comment:
When there is a single line in the if statement, you can drop the brackets, it makes the code less cluttered